magento/pwa-studio
# [PWA-2065] Add reCaptcha in checkout

## Description

Add reCaptcha in checkout on the Place Order and Braintree payment form

- Also fixes issue where reCaptcha badge was still rendering on form even if reCaptcha config was turned off in Admin.
- Also fixes issue where first reCaptcha was not initialized properly if another reCaptcha was initialized at the same time.
- Also fixes issue where "Place Order" button was enabling/disabling after click. This was also occurring on production but so quickly that it was impossible to see with human eyes.

**ReCaptcha must be activated in the backend for Place Order and Braintree payment form**

Back-end pull-request : https://github.com/magento-commerce/magento2-pwa/pull/24

A fully functional test environment is available here : https://jeremycharron-o4saqja-4txwfa4z2sor6.eu-4.magentosite.cloud/
Please get in touch with me if you need an Admin account.

## Related Issue

Closes PWA-2065

## Acceptance

### Verification Stakeholders

Project maintainers

### Specification




## Verification Steps

**ReCaptcha must be activated in the backend for Place Order and Braintree payment form**

#### Test scenario(s) for direct fix/feature


1. With reCaptcha threshold to 1.0, input credit card and click "Review order", then validate that you have a "reCaptcha validation failed error"
2. With reCaptcha threshold to 0.0, input credit card and click "Review order", then validate that credit card was succesfully added
3. With reCaptcha threshold to 1.0, click on "Place order" then validate that you have a "reCaptcha validation failed error"
4. With reCaptcha threshold to 0.0, click on "Place order" then validate that order was successful
5. Make sure that X-Captcha token is passed in the header
6. Test new Cypress test `googleReCaptcha/verifyCheckoutReCaptcha.spec.js`
7. Make sure that multiple reCaptchas in checkout are working (placeOrder + braintree + signIn)

#### Test scenario(s) for any existing impacted features/areas

1. Run Jest tests
2. Run Cypress test `googleReCaptcha/verifyForgetPasswordReCaptcha.spec.js` (was renamed)
3. Test all other reCaptcha (forget password, create account, login) to make sure they still work

#### Test scenario(s) for any Magento Backend Supported Configurations

N/A

#### Is Browser/Device testing needed?

Yes (Chrome, Safari, Firefox, Edge) + mobile

#### Any ad-hoc/edge case scenarios that need to be considered?

No

## Screenshots / Screen Captures (if appropriate)

## Breaking Changes (if any)

No breaking change

## Checklist



-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.
